src/docs/hidden-classes.md

```diff
@@ -0,0 +1,233 @@
+---
+title: 'Maps (Hidden Classes) in V8'
+description: 'How does V8 track and optimize the perceived structure of your objects?'
+---
+
+Let's show how V8 builds it's hidden classes. The principal data structures are:
+
+- **Map**: the hidden class itself. It's the first pointer value in an object and therefore allows easy comparison to see if two objects have the same class.
+- **DescriptorArray**: The full list of properties that this class has along with information about them. In some cases, the property value is even in this array.
+- **TransitionArray**: An array of "edges" from this `Map` to sibling Maps. Each edge is a property name, and should be thought of as "if I were to add a property with this name to the current class, what class would I transition to?"
+
+Because many `Map` objects only have one transition to another one (ie, they are "transitional" maps, only used on the way to something else), V8 doesn't always create a full-blown `TransitionArray` for it. Instead it'll just link directly to this "next" `Map`. The system has to do a bit of spelunking in the `DescriptorArray` of the `Map` being pointed to in order to figure out the name attached to the transition.
+
+This is an extremely rich subject. It's also subject to change, though, if you understand the concepts in this article future change should be incrementally understandable.
+
+## Why have hidden classes?
+
+V8 could do without hidden classes, sure. It would treat each object as a bag of properties. However, a very useful principle would have been left laying around: the principal of intelligent design. V8 surmises that you'll only create so many **different** kinds of objects. And each kind of object will be used in what can eventually be seen to be stereotypical ways. I say "eventually be seen" because the JavaScript language is a scripting language, not a pre-compiled one. So V8 never knows what will come next. To make use of intelligent design (that is, the assumption that there is a mind behind the code coming in), V8 has to watch and wait, letting the sense of structure seep in. The hidden class mechanism is the principal means to do this. Of course, it presupposes a sophisticated listening mechanism, and these are the Inline Caches (ICs) about which much has been written.
+
+So, if you are convinced this is good and necessary work, follow me!
+
+## An example
+
+```javascript
+function Peak(name, height, extra) {
+  this.name = name;
+  this.height = height;
+  if (isNaN(extra)) {
+    this.experience = extra;
+  } else {
+    this.prominence = extra;
+  }
+}
+
+m1 = new Peak("Matterhorn", 4478, 1040);
+m2 = new Peak("Wendelstein", 1838, "good");
+```
+
+With this code we've already got an interesting map tree from the root map (also known as the initial map) which is attached to the function `Peak`:
+
+<figure>
+  <img src="/_img/docs/hidden-classes/drawing-one.svg" width="400" height="480" alt="" loading="lazy">
```

Comment:
Please add `alt`s to the images. We probably don't need full `figcaption`s, because the following description explains what is what in the image, but at least `alt` for people who can't see images could be helpful.